backend/de.metas.contracts/src/main/java/de/metas/contracts/repository/impl/SubscriptionDiscountRepository.java

```diff
@@ -0,0 +1,121 @@
+import de.metas.contracts.SubscriptionDiscountLineId;
+import de.metas.contracts.commission.model.I_C_Flatrate_Conditions;
+import de.metas.contracts.model.I_C_SubscrDiscount;
+import de.metas.contracts.model.I_C_SubscrDiscount_Line;
+import de.metas.contracts.model.X_C_SubscrDiscount_Line;
+import de.metas.contracts.repository.ISubscriptionDiscountRepository;
+import de.metas.contracts.repository.SubscriptionDiscountQuery;
+import de.metas.product.model.I_M_Product;
+import de.metas.util.Services;
+import de.metas.util.lang.Percent;
+import org.adempiere.ad.dao.IQueryBL;
+import org.adempiere.ad.dao.IQueryBuilder;
+import org.adempiere.ad.dao.impl.CompareQueryFilter;
+import org.compiere.model.IQuery;
+import org.springframework.stereotype.Repository;
+
+import javax.annotation.Nullable;
+import java.util.Calendar;
+import java.util.Objects;
+
+@Repository
+public class SubscriptionDiscountRepository implements ISubscriptionDiscountRepository
+{
+	final private IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	@Nullable
+	public SubscriptionDiscount getDiscountOrNull(final SubscriptionDiscountQuery query)
+	{
+
+		final Calendar instance = Calendar.getInstance();
+		instance.setTimeInMillis(query.getOnDate().getTime());
+		final int month = 1 + instance.get(Calendar.MONTH);
+		final int dayOfMonth = 1 + instance.get(Calendar.DAY_OF_MONTH);
+		final IQueryBuilder<I_C_SubscrDiscount> subscrDiscountIdQuery = queryBL.createQueryBuilder(I_C_Flatrate_Conditions.class)
+				.addEqualsFilter(I_C_Flatrate_Conditions.COLUMNNAME_C_Flatrate_Conditions_ID, query.getFlatrateConditionId())
+				.andCollect(I_C_Flatrate_Conditions.COLUMN_C_SubscrDiscount_ID, I_C_SubscrDiscount.class);
+
+		final IQuery<I_M_Product> productQuery = queryBL.createQueryBuilder(I_M_Product.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_M_Product.COLUMNNAME_M_Product_ID, query.getProductId())
+				.create();
+
+		final I_C_SubscrDiscount_Line lowestSeqNoLine = queryBL.createQueryBuilder(I_C_SubscrDiscount_Line.class)
+				.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_C_SubscrDiscount_ID, I_C_Flatrate_Conditions.COLUMNNAME_C_SubscrDiscount_ID, subscrDiscountIdQuery.create())
+				.addOnlyActiveRecordsFilter()
+				.filter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_ID, null)
+								.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_ID, I_M_Product.COLUMNNAME_M_Product_ID, productQuery)
+								.setJoinOr()
+						)
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_Category_ID, null)
+								.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_Category_ID, I_M_Product.COLUMNNAME_M_Product_Category_ID, productQuery)
```

Comment:
u can use de.metas.product.IProductDAO#retrieveProductAndCategoryIdByProductId to get the Product category of that product.
And then u can filter by it.

Reply:
yup, rewrote.